tonic/functional/event_downsampling.py

```diff
@@ -0,0 +1,163 @@
+    
+    names = ["x", "y", "p", "t"]
+    formats = ['i4', 'i4', 'i4', 'i4']
+    
+    dtype = np.dtype({'names': names, 'formats': formats})
+    
+    return unstructured_to_structured(events_new.copy(), dtype=dtype)
+    
+def integrator_downsample(events: np.ndarray, sensor_size: tuple, target_size: tuple, dt: float, noise_threshold: int = 0, 
+                          differentiator_call: bool = False):
+    """Spatio-temporally downsample using with the following steps:
+    
+    1. Differencing of ON and OFF events to counter camera shake or jerk.
+    2. Use an integrate-and-fire (I-F) neuron model with a noise threshold similar to 
+    the membrane potential threshold in the I-F model to eliminate high-frequency noise.
+    
+    Multiply x/y values by a spatial_factor obtained by dividing sensor size by the target size.
+    
+    Parameters:
+        events (ndarray): ndarray of shape [num_events, num_event_channels].
+        sensor_size (tuple): a 3-tuple of x,y,p for sensor_size.
+        target_size (tuple): a 2-tuple of x,y denoting new down-sampled size for events to be
+                             re-scaled to (new_width, new_height).
+        dt (float): temporal resolution of events in milliseconds.
+        noise_threshold (int): number of events before a spike representing a new event is emitted.
+        differentiator_call (bool): Preserve frame spikes for differentiator method in order to optimise 
+                                    differentiator method.
+        
+    Returns:
+        the spatio-temporally downsampled input events using the integrator method.
+    """
+    
+    assert "x" and "y" and "t" in events.dtype.names
+    assert isinstance(noise_threshold, int)
+    assert dt is not None
+    
+    events = events.copy()
+    
+    if np.issubdtype(events["t"].dtype, np.integer):
+        dt *= 1000
+        dt_scaling = True
+    
+    if differentiator_call:
+        assert dt // events["t"][-1] == 0
+    
+    # Downsample
+    spatial_factor = np.asarray(target_size) / sensor_size[:-1]
+
+    events["x"] = events["x"] * spatial_factor[0]
+    events["y"] = events["y"] * spatial_factor[1]
+    
+    # Re-format event times to new temporal resolution
+    events_sliced = slice_events_by_time(events, time_window=dt)
+    
+    # Running buffer of events in each pixel
+    frame_spike = np.zeros(np.flip(target_size))
+    event_histogram = []
+    
+    events_new = []
+    
+    for time, event in enumerate(events_sliced):
+        # Separate by polarity
+        xy_pos = event[event["p"] == 1]
+        xy_neg = event[event["p"] == 0]
+        
+        # Sum in 2D space using histogram
+        frame_histogram = np.subtract(to_frame_numpy(xy_pos, sensor_size=(*target_size, 1), n_event_bins=1), 
+                                      to_frame_numpy(xy_neg, sensor_size=(*target_size, 1), n_event_bins=1))[0,0,...]
```

Comment:
this part here, since anyway you're working with frame_histograms, can you not just simply call 
```
all_frame_histograms = to_frame_numpy(xy_pos, sensor_size=(*target_size, 2), time_window=dt)
frame_histogram_diffs = all_frame_histograms[:, 1] -  all_frame_histograms[:, 0]

for frame_histogram in frame_histogram_diffs:
... do your thing
```
That would compute them in one go.

Reply:
Thank you for the suggestion. I've added this.